### PR TITLE
Distinguish key not found vs error accessing store

### PR DESCRIPTION
**Pull request by vpandey-RH**
Fixes [#658](https://github.com/gluster/glusterd2/issues/658). key not found and error accessing store and return 404 and 503 respectively in all user facing utility functions like GetVolume, GetPeer etc.

Signed-off-by: Vishal Pandey <vpandey@redhat.com>

### CODE REVIEW

**Comment by vpandey-RH** [comment]
@prashanthpai Kindly review the PR and check if I have missed any function that needs to be handled too.

File: glusterd2/commands/peers/addpeer.go [hunk 1]
```diff
@@ -80,7 +80,11 @@ func addPeerHandler(w http.ResponseWriter, r *http.Request) {
 	// Get the new peer information to reply back with
 	newpeer, err := peer.GetPeer(rsp.PeerID)
 	if err != nil {
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "new peer was added, but could not find peer in store. Try again later.")
+		if err == errors.ErrPeerNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/peers/addpeer.go]: This isn't valid change here. Peer not found is an incorrect response to return to client for an add peer operation.

File: glusterd2/commands/volumes/volume-create.go [hunk 2]
```diff
@@ -140,7 +140,11 @@ func volumeCreateHandler(w http.ResponseWriter, r *http.Request) {
 		// FIXME: If volume was created successfully in the txn above and
 		// then the store goes down by the time we reach here, what do
 		// we return to the client ?
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		if err == gderrors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-create.go]: Same comment as above here. Not applicable in this situation.

File: glusterd2/commands/volumes/volume-delete.go [hunk 3]
```diff
@@ -83,7 +83,11 @@ func volumeDeleteHandler(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		// TODO: Distinguish between volume not present (404) and
 		// store access failure (503)
-		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		if err == errors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-delete.go]: Now that you've fixed this, remove the above comment, here and everywhere else.

File: glusterd2/commands/volumes/volume-expand.go [hunk 4]
```diff
@@ -126,7 +126,11 @@ func volumeExpandHandler(w http.ResponseWriter, r *http.Request) {
 
 	volinfo, err := volume.GetVolume(volname)
 	if err != nil {
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		if err == errors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-expand.go]: Again, not applicable here.

File: glusterd2/commands/volumes/volume-statedump.go [hunk 5]
```diff
@@ -118,7 +118,11 @@ func volumeStatedumpHandler(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		// TODO: Distinguish between volume not present (404) and
 		// store access failure (503)
-		restutils.SendHTTPError(ctx, w, http.StatusNotFound, gderrors.ErrVolNotFound)
+		if err == gderrors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-statedump.go]: Remove above comment as you've fixed the pending TODO.

File: glusterd2/commands/volumes/volume-status.go [hunk 6]
```diff
@@ -22,7 +22,11 @@ func volumeStatusHandler(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		// TODO: Distinguish between volume not present (404) and
 		// store access failure (503)
-		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		if err == errors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-status.go]: Remove above comment as you've fixed the pending TODO.

File: glusterd2/commands/volumes/volume-stop.go [hunk 7]
```diff
@@ -93,7 +93,11 @@ func volumeStopHandler(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
 		// TODO: Distinguish between volume not present (404) and
 		// store access failure (503)
-		restutils.SendHTTPError(ctx, w, http.StatusNotFound, errors.ErrVolNotFound)
+		if err == errors.ErrVolNotFound {
```
Comment (prashanthpai) [inline glusterd2/commands/volumes/volume-stop.go]: Remove above comment as you've fixed the pending TODO.

File: glusterd2/volume/volume-utils.go [hunk 8]
```diff
@@ -147,7 +147,7 @@ func GetBrickMountDevice(brickPath, mountRoot string) (string, error) {
 			return entry.fsName, nil
 		}
 	}
-	return "", errors.New("Mount Point Not Found")
+	return "", gderrors.ErrMountPtNotFound
```
Comment (prashanthpai) [inline glusterd2/volume/volume-utils.go]: As this error is not used anywhere else, doesn't have to be an error constant.

File: glusterd2/daemon/store.go [hunk 9]
```diff
@@ -49,7 +48,7 @@ func getDaemon(id string) (Daemon, error) {
 	}
 
 	if resp.Count != 1 {
-		return nil, errors.New("daemon not found")
+		return nil, gderrors.ErrDaemonNotFound
```
Comment (prashanthpai) [inline glusterd2/daemon/store.go]: As this error is not used anywhere else, doesn't have to be an error constant.

File: plugins/device/rest.go [hunk 10]
```diff
@@ -69,7 +73,11 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
 	peerInfo, err = peer.GetPeer(peerID)
 	if err != nil {
 		logger.WithError(err).Error("Failed to get peer from store")
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Failed to get peer from store")
+		if err == errors.ErrPeerNotFound {
```
Comment (prashanthpai) [inline plugins/device/rest.go]: This is not a valid change.

File: plugins/georeplication/rest.go [hunk 11]
```diff
@@ -1033,7 +1061,11 @@ func georepSSHKeyGenerateHandler(w http.ResponseWriter, r *http.Request) {
 
 	sshkeys, err := getSSHPublicKeys(volname)
 	if err != nil {
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		if err == errors.ErrPublicSSHKeyNotFound {
```
Comment (prashanthpai) [inline plugins/georeplication/rest.go]: Nope. Not here.

File: plugins/georeplication/rest.go [hunk 12]
```diff
@@ -1050,11 +1082,15 @@ func georepSSHKeyGetHandler(w http.ResponseWriter, r *http.Request) {
 
 	sshkeys, err := getSSHPublicKeys(volname)
 	if err != nil {
-		logger.WithFields(log.Fields{
-			"error":   err.Error(),
-			"volname": volname,
-		}).Error("failed to get SSH public Keys")
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, err)
+		if err == errors.ErrPublicSSHKeyNotFound {
```
Comment (prashanthpai) [inline plugins/georeplication/rest.go]: You're really over-doing it ;)

File: plugins/georeplication/store-utils.go [hunk 13]
```diff
@@ -118,7 +118,7 @@ func getSSHPublicKeys(volname string) ([]georepapi.GeorepSSHPublicKey, error) {
 	}
 
 	if resp.Count < 1 {
-		return nil, errors.New("SSH Public Keys not found")
+		return nil, gderrors.ErrPublicSSHKeyNotFound
```
Comment (prashanthpai) [inline plugins/georeplication/store-utils.go]: As this error is not used anywhere else, doesn't have to be an error constant.

Comment (vpandey-RH) [reply plugins/georeplication/store-utils.go]: Just to mantain consistency in how we are handling errors, I thought of adding ErrPublicSSHKeyNotFound and ErrDaemonNotFound

File: plugins/georeplication/rest.go (hunk 11 above)
Comment (vpandey-RH) [reply plugins/georeplication/rest.go]: getSSHPublicKeys() handles two errors in its definition same as GetVolume so I guess we can include this error as well in this PR.

Comment (prashanthpai) [reply plugins/georeplication/rest.go]: I don't prefer this PR to change add new error variables if it's used only once.

File: plugins/device/rest.go [hunk 14]
```diff
@@ -68,8 +72,11 @@ func deviceAddHandler(w http.ResponseWriter, r *http.Request) {
 	}
 	peerInfo, err = peer.GetPeer(peerID)
 	if err != nil {
-		logger.WithError(err).Error("Failed to get peer from store")
-		restutils.SendHTTPError(ctx, w, http.StatusInternalServerError, "Failed to get peer from store")
+		if err == errors.ErrPeerNotFound {
```
Comment (prashanthpai) [inline plugins/device/rest.go]: Shouldn't be done here.

**Comment by prashanthpai** [comment]
Also, needs a rebase.

**Comment by prashanthpai** [comment]
retest this please

**Pull request merged**